controllers/config/clientconfig_controller.go

```diff
@@ -0,0 +1,206 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	configapi "github.com/k8ssandra/k8ssandra-operator/apis/config/v1beta1"
+	k8ssandraapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/clientcache"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+)
+
+const (
+	ClientConfigHashAnnotation = k8ssandraapi.ResourceHashAnnotation
+	KubeSecretHashAnnotation   = "k8ssandra.io/secret-hash"
+)
+
+type ClientConfigReconciler struct {
+	Scheme       *runtime.Scheme
+	ClientCache  *clientcache.ClientCache
+	shutdownFunc context.CancelFunc
+
+	// filterMutex  sync.RWMutex
+	secretFilter map[types.NamespacedName]types.NamespacedName
+}
+
+/*
+	What about situations:
+		- Remove clientconfig when the cluster is still added?
+			* Cluster properties get updated?
+				* Do we accept or not?
+			* Could just be a cert-update etc, short-term, long-term ones
+		- New ClientConfig doesn't work?
+		- Add cluster points to a clientconfig that doesn't exist?
+		- What if the ClientConfig's kubeConfig is changed in the secret?
```

Comment:
In the context of this PR I am inclined to say that it should trigger a restart. Maybe a delayed restart where the delay is configurable. I suggest a delay in case there are multiple ClientConfigs to be updated.

Reply:
It does trigger a restart in this PR's implementation.

Reply:
What do you think about a delayed restart?

Reply:
What would be the benefit of delayed restart?

Reply:
Let's say we create 3 ClientConfig objects. I haven't tested yet but I suspect that would result in 3 successive restarts.